### PR TITLE
Add selection to debrief session

### PR DESCRIPTION
**Pull request by r-gaia-cs**
Address #213.

**This doesn't have a link from the home page.**

### CODE REVIEW

**Comment by wking** [comment]
I'm still uncertain about per-task views [1], but other than that this
looks reasonable to me.  I don't think the trailing comma is legal
though [2].  You should probably use:

  <a href="mailto:{% for task in all_tasks %}{{ task.person.email }}{% if not last %},{% endif %}{% endfor %}">…</a>

[1]: https://github.com/swcarpentry/amy/issues/213#issuecomment-77088428
[2]: https://tools.ietf.org/html/rfc6068#section-2
[3]: https://docs.djangoproject.com/en/1.7/ref/templates/builtins/#for

**Comment by wking** [comment]
Also, it's better to push as much of the filtering as possible down
into the database, to save the cost of converting everything into
Python objects.  Maybe:

  tasks = Task.objects.filter(
      event__end__gte=form.cleaned_data['begin_date'],
      event__begin__lte=form.cleaned_data['end_date'],
      role__name='instructor',
      ).order_by('event', 'person', 'role')

for your whole form.is_valid() block ;).

We don't want to mail instructors who have disabled access (see #161).
Do we want that filtering to happen in the view, or in the template
while constructing the mailto?  We might also have instructors with
NULL email (see #125, 4659c06, and #140).  Do we filter *them* out in
the view or the template?

**Comment by r-gaia-cs** [comment]
@wking Thanks for the comments.

> Also, it's better to push as much of the filtering as possible down
> into the database, to save the cost of converting everything into
> Python objects

I got "Relation fields do not support nested lookups" when trying your suggestion.

> Do we want that filtering to happen in the view, or in the template
> while constructing the mailto? (...) Do we filter *them* out in
> the view or the template?

You can vote on this. =)

**Comment by wking** [comment]
On Fri, Mar 06, 2015 at 06:42:02AM -0800, Raniere Silva wrote:
> > Also, it's better to push as much of the filtering as possible
> > down into the database, to save the cost of converting everything
> > into Python objects
>
> I got "Relation fields do not support nested lookups" when trying
> your suggestion.

Ah, looks like it's Event.start [1], so:

  tasks = Task.objects.filter(
      event__end__gte=form.cleaned_data['begin_date'],
      event__start__lte=form.cleaned_data['end_date'],
      role__name='instructor',
      ).order_by('event', 'person', 'role') 

> > Do we want that filtering to happen in the view, or in the
> > template while constructing the mailto? (...) Do we filter *them*
> > out in the view or the template?
> 
> You can vote on this. =)

I'm happy either way ;).  It's just a question of whether or not users
(SWC admins) will want to know about the matching instructors that
they can't contact (in which case we should filter in the mailto) or
if they only care about folks they can contact (in which case we
should filter in the view).  So if you're contacting instructors for
debriefing, do you want to see entries in the results for instructors
you can't contact?

[1]: https://github.com/swcarpentry/amy/blob/546004032019ecc57e8834280d5bbad6be233d1c/workshops/models.py#L256

**Comment by r-gaia-cs** [comment]
> It's just a question of whether or not users
> (SWC admins) will want to know about the matching instructors that
> they can't contact (in which case we should filter in the mailto) or
> if they only care about folks they can contact (in which case we
> should filter in the view).  So if you're contacting instructors for
> debriefing, do you want to see entries in the results for instructors
> you can't contact?

I update this using @wking suggest to use `filter` and removing instructors that we can't contact.
Instructors without email or blank email are listed but not include in the "send email" link.

@wking Could you check this using as input `2015-07-01` and `2015-07-30`? Jinja's `last` variable isn't working.

**Comment by wking** [comment]
On Fri, Mar 06, 2015 at 11:29:56AM -0800, Raniere Silva wrote:
> Jinja's `last` variable isn't working.

Ah, looks like you need the namespace mentioned in the docs I'd linked
[1].  Try:

  {% if not forloop.last %}

Django's template language is very similar to Jinja, but they're not
quite the same.

[1]: https://docs.djangoproject.com/en/1.7/ref/templates/builtins/#for

**Comment by r-gaia-cs** [comment]
> Ah, looks like you need the namespace

Thanks. Fixed.

File: workshops/forms.py [hunk 1]
```diff
@@ -71,3 +71,11 @@ class SearchForm(forms.Form):
     in_persons = forms.BooleanField(label='in persons',
                                     required=False,
                                     initial=True)
+
+class DebriefForm(forms.Form):
+    '''Represent general debrief form.'''
+
+    begin_date = forms.CharField(label='Begin date as YYYY-MM-DD',
+                           max_length=10)
+    end_date = forms.CharField(label='End date as YYYY-MD-DD',
```
Comment (pbanaszkiewicz) [inline workshops/forms.py]: Little typo; should be "End date as YYYY-MM-DD". I'd also add a `min_length=10` parameter to help with validation. Or do we accept formats like "YYYY-MM"?

(We can go pro and use a custom validator https://docs.djangoproject.com/en/1.7/ref/validators/#writing-validators to make sure a valid date is provided. But I don't see it as a necessity.)

**Comment by pbanaszkiewicz** [comment]
@r-gaia-cs: Looks good to me, although I think we need a link to `/debrief/` subpage somewhere.

File: workshops/forms.py (hunk 1 above)
Comment (wking) [inline workshops/forms.py]: On Sun, Mar 08, 2015 at 12:36:00PM -0700, Piotr Banaszkiewicz wrote:
> We can go pro and use a custom validator
> https://docs.djangoproject.com/en/1.7/ref/validators/#writing-validators
> to make sure a valid date is provided. But I don't see it as a
> necessity.

I've also proposed Bootstrap's datepicker (see #110) for handling date
input.  It's harder to break your input if you're just clicking on
calendar entries.

Comment (pbanaszkiewicz) [inline workshops/forms.py]: > I've also proposed Bootstrap's datepicker (see #110) for handling date input. It's harder to break your input if you're just clicking on calendar entries.

It's a good UX idea, but if @r-gaia-cs doesn't want to include this widget we can create an issue for this.

**Comment by r-gaia-cs** [comment]
@pbanaszkiewicz Can you add the link to `/debrief/` and change this to use Bootstrap's datepicker (see #110)?

**Comment by pbanaszkiewicz** [comment]
Okay. I think both link to `/debrief/` and Bootstrap's datepicker should land after #221. I'll make an issue for that.

**Pull request merged**